File: 200-number-of-islands/number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        rows, cols = len(grid), len(grid[0])
        visited = set()
        numIslands = 0

        def dfs(r, c):
            if r < 0 or c < 0 or r >= rows or c >= cols or grid[r][c] == "0" or (r, c) in visited:
                return

            visited.add((r, c))

            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c - 1)
            dfs(r, c + 1)


        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r, c) not in visited:
                    numIslands += 1
                    dfs(r, c)

        return numIslands            


                

        

File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        Set<String> seen = new HashSet<>();
        int count = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                  if (grid[i][j] == '1') {
                      dfs(grid, i, j);
                      count++;  
                  }
            }    
        }
        return count;
    }

    private void dfs(char[][] grid, int row, int col) {
        int rows = grid.length;
        int cols = grid[0].length;
        //System.out.println("row:"+row + "col:"+col);

        if (row < 0 || row >= rows || col < 0|| col >= cols || grid[row][col] == '0') {
            return;
        }
        grid[row][col]= '0';
        dfs(grid, row + 1, col);
        dfs(grid, row - 1, col);
        dfs(grid, row, col + 1);
        dfs(grid, row, col - 1);
    }
}